### ISSUE TITLE
STAT vs LSTAT in SFTP mode on proftpd

### ISSUE THREAD

**Issue by konung** [proftpd/proftpd#1647]
### What I Did

We have a working ProFTPd server running SFTP/FTP. FileZilla, bash and various ruby/python sftp scripts are connecting just fine. 

However, we have one JS client, that we don't control, that's trying to connect to us, and it is running into problems(it's NetSuite JS script).  

Based on logs, I determined that when JS sftp client library is issuing `conn.list();` command, it actually is trying to call "STAT" on our server instead of "LSTAT" like all the other sftp clients. 

### What I Expected/Wanted

I was wondering if there is any directive (or other trick), that I haven't found in docs, that could alias STAT to LSTAT.  This JS library works against some other "mysterious SFTP server" . 

The sftp JS library they are using, and code is in official NetSuite/Oracle docs, so I'm guessing NetSuite tested it against something. Likely, something Windows-based, from what I gleaned from hints on various forums / Stack Overflow. 

### ProFTPD Version and Configuration

```bash
Compile-time Settings:
  Version: 1.3.6c (maint)
  Platform: LINUX [Linux 5.4.0-144-generic x86_64]
  Built: Thu Feb 27 2020 19:34:56 UTC
  Built With:
    configure  '--build=x86_64-linux-gnu' '--includedir=${prefix}/include' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--sysconfdir=/etc' '--localstatedir=/var' '--disable-silent-rules' '--libdir=${prefix}/lib/x86_64-linux-gnu' '--libexecdir=${prefix}/lib/x86_64-linux-gnu' '--disable-maintainer-mode' '--disable-dependency-tracking' '--prefix=/usr' '--with-includes=/usr/include/postgresql:/usr/include/mysql' '--mandir=/usr/share/man' '--sysconfdir=/etc/proftpd' '--localstatedir=/run' '--libexecdir=/usr/lib/proftpd' '--enable-sendfile' '--enable-facl' '--enable-dso' '--enable-autoshadow' '--enable-ctrls' '--enable-ipv6' '--enable-nls' '--enable-memcache' '--with-lastlog=/var/log/lastlog' '--enable-pcre' '--disable-strip' '--enable-redis' '--build' 'x86_64-linux-gnu' '--with-shared=mod_unique_id:mod_site_misc:mod_load:mod_ban:mod_quotatab:mod_sql:mod_sql_mysql:mod_sql_postgres:mod_sql_sqlite:mod_sql_odbc:mod_dynmasq:mod_quotatab_sql:mod_ldap:mod_quotatab_ldap:mod_ratio:mod_tls:mod_rewrite:mod_radius:mod_wrap:mod_wrap2:mod_wrap2_file:mod_wrap2_sql:mod_quotatab_file:mod_quotatab_radius:mod_facl:mod_ctrls_admin:mod_copy:mod_deflate:mod_ifversion:mod_geoip:mod_exec:mod_sftp:mod_sftp_pam:mod_sftp_sql:mod_shaper:mod_sql_passwd:mod_ifsession:mod_auth_otp:mod_tls_redis:mod_wrap2_redis:mod_redis:mod_memcache:mod_tls_memcache:mod_readme:mod_snmp' 'build_alias=x86_64-linux-gnu' 'CFLAGS=-g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-oI4YZ3/proftpd-dfsg-1.3.6c=. -fstack-protector-strong -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-Bsymbolic-functions -Wl,-z,relro' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-oI4YZ3/proftpd-dfsg-1.3.6c=. -fstack-protector-strong -Wformat -Werror=format-security'

  CFLAGS: -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-oI4YZ3/proftpd-dfsg-1.3.6c=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration
  LDFLAGS: -L$(top_srcdir)/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -rdynamic  -L/usr/lib/x86_64-linux-gnu -L/usr/lib/x86_64-linux-gnu
  LIBS: -lacl  -lpcreposix -lpcre -lssl -lcrypto -lcap  -lpam -lsupp -lattr -lnsl -lresolv -lresolv -lcrypt -ldl -lhiredis -lmemcachedutil -lmemcached  -pthread

  Files:
    Configuration File:
      /etc/proftpd/proftpd.conf
    Pid File:
      /run/proftpd.pid
    Scoreboard File:
      /run/proftpd.scoreboard
    Header Directory:
      /usr/include/proftpd
    Shared Module Directory:
      /usr/lib/proftpd

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    + Autoshadow support
    + Controls support
    + curses support
    - Developer support
    + DSO support
    + IPv6 support
    + Largefile support
    + Lastlog support
    + Memcache support
    + ncursesw support
    + NLS support
    + Redis support
    - Sodium support
    + OpenSSL support
    + PCRE support
    + POSIX ACL support
    + Shadow file support
    + Sendfile support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 4096
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10
```

 I would appriciate any pointers. 
 
 Thank you.

**Comment by Castaglia** [comment]
Is there any more information on just what the problem (as seen by this JS script/client) is?  Any error messages, `mod_sftp` logs, or anything that might help provide more context on the issue?

**Comment by konung** [comment]
Hi. 

On the client side they are using NetSuite "N/sftp" module - https://docs.oracle.com/en/cloud/saas/netsuite/ns-online-help/section_4617004932.html  . 

As I understand, they are using SuiteScript wich is a restrictive version of ECMAScript 5.1.  It's restrictive - so they can't use other packages like "ssh2-sftp-client" npm packages - which connects just fine in testing. So we know the issues is with "N/sftp" doing something weird, and sending STAT for whatever reason. 

Unfortunrely the source code is closed for "N/sftp". The logs that our vendor can produce on Netsuite just say "Unexpected Error". When they talked to NetSuite support , their backend logs are also saying "Unexpected Error". So they are investigating on their end, but so far they haven't found anything. 

Hence, I'm trying to see if there is anything I can do on Proftpd side. Maybe alias a STAT to LSTAT? Is that even a thing?


Here are the logs from our side:  I redacted the IP address(64.xxx.xxx.xxx ), and username (NETSUITEUSER)

```log
4995:2023-03-28 13:07:09,255 64.xxx.xxx.xxx ssh2 - 0 - - 0 - - - KEXINIT - KEXINIT
4997:2023-03-28 13:07:09,294 64.xxx.xxx.xxx ssh2 - 0 - - 0 - - - ECDH_INIT - ECDH_INIT
4998:2023-03-28 13:07:09,359 64.xxx.xxx.xxx ssh2 - 0 - - 0 - - - NEWKEYS - NEWKEYS
4999:2023-03-28 13:07:09,385 64.xxx.xxx.xxx ssh2 - 0 - - 0 - - - SERVICE_REQUEST - SERVICE_REQUEST
5000:2023-03-28 13:07:09,386 64.xxx.xxx.xxx ssh2 - 0 - - 0 - - - USER 331 USER NETSUITEUSER
5001:2023-03-28 13:07:09,387 64.xxx.xxx.xxx ssh2 - 0 - - 0 - - - USERAUTH_REQUEST - USERAUTH_REQUEST NETSUITEUSER none
5002:2023-03-28 13:07:09,415 64.xxx.xxx.xxx ssh2 - 0 - - 0 - - - USER 331 USER NETSUITEUSER
5003:2023-03-28 13:07:09,458 64.xxx.xxx.xxx ssh2 NETSUITEUSER 1004 - - 43 - - - PASS 230 PASS (hidden)
5004:2023-03-28 13:07:09,459 64.xxx.xxx.xxx ssh2 NETSUITEUSER 1004 - - 0 - - - USERAUTH_REQUEST - USERAUTH_REQUEST NETSUITEUSER password
5006:2023-03-28 13:07:09,485 64.xxx.xxx.xxx ssh2 NETSUITEUSER 1004 - - 0 - - - CHANNEL_OPEN - CHANNEL_OPEN session
5009:2023-03-28 13:07:09,513 64.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 0 - - - CHANNEL_REQUEST - CHANNEL_REQUEST subsystem
5011:2023-03-28 13:07:09,540 64.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 0 - - - INIT - INIT 6
5013:2023-03-28 13:07:09,569 64.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 1 - - - STAT - STAT ./from_customer/
5015:2023-03-28 13:07:09,597 64.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 0 - - - CHANNEL_EOF - CHANNEL_EOF
5016:2023-03-28 13:07:09,598 64.xxx.xxx.xxx ssh2 NETSUITEUSER 1004 - - 0 - - - CHANNEL_CLOSE - CHANNEL_CLOSE
```


The error is happening on line "5013:"  when our Proftpd server get's that command.  Netsuite client just errors out. 


Here is a working connection log from Vultr test box to our firewalled Proftpd server, using the same credentials and "ssh-sftp-client" ( when I tried Mac CLI sftp tool - log looks the same for the same commands) . I'm testing from external box to eliminated firewall, networking issues, etc. 

```log
2023-03-28 13:23:06,940 45.xxx.xxx.xxx ssh2 - 0 - - 0 - - - KEXINIT - KEXINIT  
2023-03-28 13:23:06,941 45.xxx.xxx.xxx ssh2 - 0 - - 0 - - - ECDH_INIT - ECDH_INIT  
2023-03-28 13:23:06,954 45.xxx.xxx.xxx ssh2 - 0 - - 0 - - - NEWKEYS - NEWKEYS  
2023-03-28 13:23:06,966 45.xxx.xxx.xxx ssh2 - 0 - - 0 - - - SERVICE_REQUEST - SERVICE_REQUEST  
2023-03-28 13:23:06,968 45.xxx.xxx.xxx ssh2 - 0 - - 0 - - - USER 331 USER NETSUITEUSER  
2023-03-28 13:23:06,968 45.xxx.xxx.xxx ssh2 - 0 - - 0 - - - USERAUTH_REQUEST - USERAUTH_REQUEST NETSUITEUSER none  
2023-03-28 13:24:39,384 45.xxx.xxx.xxx ssh2 - 0 - - 0 - - - KEXINIT - KEXINIT  
2023-03-28 13:24:39,386 45.xxx.xxx.xxx ssh2 - 0 - - 0 - - - ECDH_INIT - ECDH_INIT  
2023-03-28 13:24:39,399 45.xxx.xxx.xxx ssh2 - 0 - - 0 - - - NEWKEYS - NEWKEYS  
2023-03-28 13:24:39,418 45.xxx.xxx.xxx ssh2 - 0 - - 0 - - - SERVICE_REQUEST - SERVICE_REQUEST  
2023-03-28 13:24:39,420 45.xxx.xxx.xxx ssh2 - 0 - - 0 - - - USER 331 USER NETSUITEUSER  
2023-03-28 13:24:39,420 45.xxx.xxx.xxx ssh2 - 0 - - 0 - - - USERAUTH_REQUEST - USERAUTH_REQUEST NETSUITEUSER none  
2023-03-28 13:24:41,040 45.xxx.xxx.xxx ssh2 - 0 - - 1 - - - USER 331 USER NETSUITEUSER  
2023-03-28 13:24:41,071 45.xxx.xxx.xxx ssh2 NETSUITEUSER 1004 - - 31 - - - PASS 230 PASS (hidden)  
2023-03-28 13:24:41,071 45.xxx.xxx.xxx ssh2 NETSUITEUSER 1004 - - 0 - - - USERAUTH_REQUEST - USERAUTH_REQUEST NETSUITEUSER password  
2023-03-28 13:24:41,072 45.xxx.xxx.xxx ssh2 NETSUITEUSER 1004 - - 0 - - - CHANNEL_OPEN - CHANNEL_OPEN session  
2023-03-28 13:24:41,073 45.xxx.xxx.xxx ssh2 NETSUITEUSER 1004 - - 0 - - - CHANNEL_REQUEST - CHANNEL_REQUEST env  
2023-03-28 13:24:41,074 45.xxx.xxx.xxx ssh2 NETSUITEUSER 1004 - - 0 - - - CHANNEL_REQUEST - CHANNEL_REQUEST env  
2023-03-28 13:24:41,074 45.xxx.xxx.xxx ssh2 NETSUITEUSER 1004 - - 0 - - - CHANNEL_REQUEST - CHANNEL_REQUEST env  
2023-03-28 13:24:41,075 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 0 - - - CHANNEL_REQUEST - CHANNEL_REQUEST subsystem  
2023-03-28 13:24:41,076 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 0 - - - INIT - INIT 3  
2023-03-28 13:24:41,078 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 1 - - - REALPATH - REALPATH .  
2023-03-28 13:24:45,408 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 1 0.000 - - OPENDIR - OPENDIR /  
2023-03-28 13:24:45,410 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 0 0.002 - - READDIR - READDIR 373f35eb6888e05a  
2023-03-28 13:24:45,412 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 0 0.004 - - READDIR 1 READDIR 373f35eb6888e05a  
2023-03-28 13:24:45,413 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 /home/apiftp/customer/ - 0 0.005 success - MLSD - MLSD /  
2023-03-28 13:24:45,413 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 0 - - - CLOSE 0 CLOSE 373f35eb6888e05a  
2023-03-28 13:24:54,864 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 0 - - - LSTAT - LSTAT /.  
2023-03-28 13:24:54,866 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 1 0.000 - - OPENDIR - OPENDIR /./  
2023-03-28 13:24:54,868 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 0 0.002 - - READDIR - READDIR 165f3523a864b94d  
2023-03-28 13:24:54,870 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 0 0.004 - - READDIR 1 READDIR 165f3523a864b94d  
2023-03-28 13:24:54,871 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 /home/apiftp/customer/ - 0 0.005 success - MLSD - MLSD /  
2023-03-28 13:24:54,871 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 0 - - - CLOSE 0 CLOSE 165f3523a864b94d  
2023-03-28 13:25:03,688 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 1 - - - LSTAT - LSTAT /from_customer  
2023-03-28 13:25:03,690 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 0 0.000 - - OPENDIR - OPENDIR /from_customer/  
2023-03-28 13:25:03,692 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 0 0.002 - - READDIR - READDIR b74c8ccc664d777a  
2023-03-28 13:25:03,694 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 0 0.004 - - READDIR 1 READDIR b74c8ccc664d777a  
2023-03-28 13:25:03,697 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 /home/apiftp/customer/from_customer - 2 0.007 success - MLSD - MLSD /from_customer  
2023-03-28 13:25:03,697 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 0 - - - CLOSE 0 CLOSE b74c8ccc664d777a  
2023-03-28 13:25:08,672 45.xxx.xxx.xxx sftp NETSUITEUSER 1004 - - 1 - - - CHANNEL_EOF - CHANNEL_EOF  
2023-03-28 13:25:08,673 45.xxx.xxx.xxx ssh2 NETSUITEUSER 1004 - - 0 - - - CHANNEL_CLOSE - CHANNEL_CLOSE  
```

@Castaglia  I would appriciate any insight.  Thank you

**Comment by konung** [comment]
Since this "N/sftp" module is the same version for multiple Netsuite/Oracle customers per our vendor, they have it communicating with anothe customer. But they don't have details about the server the other site is using.

**Comment by Castaglia** [comment]
Can you provide the ProFTPD configuration that you're using?  That might help me see how `mod_sftp` might return an error unexpectedly for these `STAT` requests.

**Comment by konung** [comment]
@Castaglia 
Hi

1. Per my last comment,  I think the “vendor” wasn't entirely “forthcoming” or there was miscommunication – because they couldn't show us how another customer using the same module is working. 

2. Unexpectedly, NetSuite escalated our problem and pushed out a hotfix for it. The issue was indeed their old "N/sftp" module – some kind of custom thing they had written in their flavor of JavaScript called SuiteScript. 

We just finished testing and pushing it in production last week – it works with Proftpd sftp without having to make any custom accommodations for it on Proftpd side. 

**TLDR: In the end the problem was legacy JavaScript sftp library on the client side.** 

Thank you so much
